compiler/rustc_privacy/src/lib.rs

```diff
@@ -739,7 +739,18 @@ impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
             }
             hir::ItemKind::Impl(ref impl_) => {
                 for impl_item_ref in impl_.items {
-                    self.update(impl_item_ref.id.owner_id.def_id, item_ev, Level::Direct);
+                    if impl_.of_trait.is_some()
+                        && !self.tcx.visibility(impl_item_ref.id.owner_id).is_public()
+                    {
+                        self.update_eff_vis(
+                            impl_item_ref.id.owner_id.def_id,
+                            item_ev,
+                            None,
+                            Level::Direct,
+                        );
+                    } else {
+                        self.update(impl_item_ref.id.owner_id.def_id, item_ev, Level::Direct);
+                    }
```

Comment:
I'm still not sure how exactly, but monomorphization collector (compiler\rustc_monomorphize\src\collector.rs) requires that if the trait impl is reachable, then its items should be reachable as well.
So the condition here should be
```rust
let nominal_vis = if impl_.of_trait.is_none().then(|| self.tcx.local_visibility(def_id));
self.update_eff_vis(impl_item_ref.id.owner_id.def_id, item_ev, nominal_vis, Level::Direct);
```